### PR TITLE
Add an external hash to modules in summaries

### PR DESCRIPTION
**Pull request by floitsch**
Two modules (libraries) are externally equal if another library importing it doesn't see any changes.

Let the compiler compute the external equality hash, instead of running through the summary.

### CODE REVIEW

File: src/compiler/lsp/protocol_summary.cc [hunk 1]
```diff
@@ -259,12 +264,51 @@ class Writer {
...
+
+  /// A version of 'printf' that keeps track of the data for the external sha1.
+  /// Any data that represents a module's external representation needs to go
+  /// through the sha1 so that we know when to recompute modules that depend on
+  /// the current module.
+  void printf_external(const char* format, ...) {
+    va_list arguments;
+    va_start(arguments, format);
+
+    // Calculate the size of the buffer required
```
Comment (kasperl) [inline src/compiler/lsp/protocol_summary.cc:294]:
```suggestion
    // Calculate the size of the buffer required.
```

File: src/compiler/lsp/protocol_summary.cc [hunk 2]
```diff
@@ -259,12 +264,51 @@ class Writer {
...
+    va_list args_copy;
+    va_copy(args_copy, arguments);
+    int size = vsnprintf(NULL, 0, format, args_copy);
+    va_end(args_copy);
+
+    if (size < 0) FATAL("Failure to convert argument to string");
+
+    // Allocate a buffer of the required size
```
Comment (kasperl) [inline src/compiler/lsp/protocol_summary.cc:302]:
```suggestion
    // Allocate a buffer of the required size.
```

File: src/compiler/lsp/protocol_summary.cc [hunk 3]
```diff
@@ -259,12 +264,51 @@ class Writer {
...
+    char* buffer = (char*)malloc(size + 1);
+
+    // Print to the buffer
```
Comment (kasperl) [inline src/compiler/lsp/protocol_summary.cc:305]:
```suggestion
    // Print to the buffer.
```

File: src/compiler/lsp/protocol_summary.cc [hunk 4]
```diff
@@ -259,12 +264,51 @@ class Writer {
...
+    // Calculate the size of the buffer required
+    va_list args_copy;
+    va_copy(args_copy, arguments);
+    int size = vsnprintf(NULL, 0, format, args_copy);
+    va_end(args_copy);
+
+    if (size < 0) FATAL("Failure to convert argument to string");
+
+    // Allocate a buffer of the required size
+    char* buffer = (char*)malloc(size + 1);
```
Comment (kasperl) [inline src/compiler/lsp/protocol_summary.cc:303]:
```suggestion
    char* buffer = static_cast<char*>(malloc(size + 1));
```

File: src/compiler/third_party/tiny-sha1/README.toitware [hunk 5]
```diff
@@ -0,0 +1,12 @@
+Name: TinySHA1
```
Comment (kasperl) [inline src/compiler/third_party/tiny-sha1/README.toitware:1]: Maybe this file should be README.toit or README.toitlang?

File: src/compiler/lsp/protocol_summary.cc [hunk 6]
```diff
@@ -259,12 +264,51 @@ class Writer {
...
+
+    // Print to the buffer
+    vsnprintf(buffer, size + 1, format, arguments);
+    va_end(arguments);
+
+    printf("%s", buffer);
+    sha1_.processBytes(buffer, size);
```
Comment (kasperl) [inline src/compiler/lsp/protocol_summary.cc:310]:
```suggestion
    sha1_.processBytes(buffer, size);
    free(buffer);
```

File: src/compiler/third_party/tiny-sha1/TinySHA1.hpp [hunk 7]
```diff
@@ -0,0 +1,197 @@
+/*
```
Comment (kasperl) [inline src/compiler/third_party/tiny-sha1/TinySHA1.hpp:1]: It is a little bit annoying to have this in addition to src/sha1.{cc,h}.

File: src/compiler/lsp/protocol_summary.cc (hunk 1 above)
Comment (floitsch) [reply src/compiler/lsp/protocol_summary.cc]: done.

File: src/compiler/lsp/protocol_summary.cc (hunk 2 above)
Comment (floitsch) [reply src/compiler/lsp/protocol_summary.cc]: done.

File: src/compiler/lsp/protocol_summary.cc (hunk 4 above)
Comment (floitsch) [reply src/compiler/lsp/protocol_summary.cc]: used `unvoid_cast` instead.
done.

File: src/compiler/lsp/protocol_summary.cc (hunk 3 above)
Comment (floitsch) [reply src/compiler/lsp/protocol_summary.cc]: done.

File: src/compiler/lsp/protocol_summary.cc (hunk 6 above)
Comment (floitsch) [reply src/compiler/lsp/protocol_summary.cc:310]: done.

File: src/compiler/third_party/tiny-sha1/README.toitware (hunk 5 above)
Comment (floitsch) [reply src/compiler/third_party/tiny-sha1/README.toitware]: I changed it to README.toit.txt.
done.

File: src/compiler/third_party/tiny-sha1/TinySHA1.hpp (hunk 7 above)
Comment (floitsch) [reply src/compiler/third_party/tiny-sha1/TinySHA1.hpp:1]: as discussed: for now we will keep it, but we will refactor it to use the src/sha1.{cc.h}.

**Pull request merged**